### PR TITLE
crimson/osd/watch: ensure Watch/Notify lifetime

### PR DESCRIPTION
**Pull request by Matan-B**
```
[==========] 16 tests from 2 test suites ran. (14372 ms total)                                                                                                              
[  PASSED  ] 14 tests.                                                                                                                                                      
[  SKIPPED ] 2 tests, listed below:
[  SKIPPED ] LibRadosWatchNotifyECPP.WatchNotify
[  SKIPPED ] LibRadosWatchNotifyECPP.WatchNotifyTimeout
...
[==========] 11 tests from 2 test suites ran. (20451 ms total)
[  PASSED  ] 10 tests.
[  SKIPPED ] 1 test, listed below:
[  SKIPPED ] LibRadosWatchNotifyEC.WatchNotif
```

```
DEBUG 2023-05-26 23:36:36,101 [shard 0] osd - ShardServices::dispatch_context_transaction: empty transaction
Reactor stalled for 95 ms on shard 1. Backtrace: 0x45d4d 0x47328a37 0x47046135 0x470607b5 0x47060c5c 0x47060db2 0x470611fd 0x12cef 0x2a6de432fcde 0x2a6de4330d18 0x2a6de43336cf 0x2a6de4334360 0x168b86 0x46c06173 0x2a6de43345d4 0x2a6de4323b19 0x2a6de4323be5 0x2a6de43150a5 0x2a6de4316f2a 0xd61c0 0x32402 0xbd8a7 0xbd134 0xbdf7a 0x3c4cdbd7 0x3c4d6633 0x470249ea 0x47081727 0x47278455 0x47299364 0x47299939 0x46e0e063 0x81c9 0x39e72
kernel callstack:
    #0 0x55698378bbd7 in auto seastar::futurize_invoke<crimson::osd::Watch::notify_ack(unsigned long, ceph::buffer::v15_2_0::list const&)::{lambda(auto:1)#1}&, seastar::shared_ptr<crimson::osd::Notify> const&>(crimson::osd::Watch::notify_ack(unsigned long, ceph::buffer::v15_2_0::list const&)::{lambda(auto:1)#1}&, seastar::shared_ptr<crimson::osd::Notify> const&) (/usr/bin/ceph-osd+0x3c4cdbd7)
    #1 0x556983794633 in seastar::internal::do_for_each_state<std::_Rb_tree_const_iterator<seastar::shared_ptr<crimson::osd::Notify> >, crimson::osd::Watch::notify_ack(unsigned long, ceph::buffer::v15_2_0::list const&)::{lambda(auto:1)#1}>::run_and_dispose() (/usr/bin/ceph-osd+0x3c4d6633)
    #2 0x55698e2e29ea in seastar::reactor::run_tasks(seastar::reactor::task_queue&) (/usr/bin/ceph-osd+0x470249ea)
    #3 0x55698e33f727 in seastar::reactor::run_some_tasks() (/usr/bin/ceph-osd+0x47081727)
    #4 0x55698e536455 in seastar::reactor::do_run() (/usr/bin/ceph-osd+0x47278455)
    #5 0x55698e557364 in seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::{lambda()#3}::operator()() const (/usr/bin/ceph-osd+0x47299364)
    #6 0x55698e557939 in std::_Function_handler<void (), seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::{lambda()#3}>::_M_invoke(std::_Any_data const&) (/usr/bin/ceph-osd+0x47299939)
    #7 0x55698e0cc063 in seastar::posix_thread::start_routine(void*) (/usr/bin/ceph-osd+0x46e0e063)
    #8 0x7fd7281131c9 in start_thread (/lib64/libpthread.so.0+0x81c9)
    #9 0x7fd72672ee72 in __clone (/lib64/libc.so.6+0x39e72)

0x604000bc2630 is located 32 bytes inside of 48-byte region [0x604000bc2610,0x604000bc2640)
freed by thread T1 (reactor-1) here:
    #0 0x7fd72b5af36f in operator delete(void*, unsigned long) (/lib64/libasan.so.6+0xb736f)
    #1 0x5569837a910a in std::_Rb_tree<seastar::shared_ptr<crimson::osd::Notify>, seastar::shared_ptr<crimson::osd::Notify>, std::_Identity<seastar::shared_ptr<crimson::osd::Notify> >, std::less<void>, std::allocator<seastar::shared_ptr<crimson::osd::Notify> > >::_M_erase(std::_Rb_tree_node<seastar::shared_ptr<crimson::osd::Notify> >*) (/usr/bin/ceph-osd+0x3c4eb10a)

previously allocated by thread T1 (reactor-1) here:
    #0 0x7fd72b5ae307 in operator new(unsigned long) (/lib64/libasan.so.6+0xb6307)
    #1 0x5569837b7cae in std::pair<std::_Rb_tree_iterator<seastar::shared_ptr<crimson::osd::Notify> >, bool> std::_Rb_tree<seastar::shared_ptr<crimson::osd::Notify>, seastar::shared_ptr<crimson::osd::Notify>, std::_Identity<seastar::shared_ptr<crimson::osd::Notify> >, std::less<void>, std::allocator<seastar::shared_ptr<crimson::osd::Notify> > >::_M_emplace_unique<seastar::shared_ptr<crimson::osd::Notify> >(seastar::shared_ptr<crimson::osd::Notify>&&) (/usr/bin/ceph-osd+0x3c4f9cae)

Thread T1 (reactor-1) created by T0 here:
    #0 0x7fd72b5507c5 in pthread_create (/lib64/libasan.so.6+0x587c5)
    #1 0x55698e0d0346 in seastar::posix_thread::posix_thread(seastar::posix_thread::attr, std::function<void ()>) (/usr/bin/ceph-osd+0x46e12346)
    #2 0x5569917b195f  (/usr/bin/ceph-osd+0x4a4f395f)

SUMMARY: AddressSanitizer: heap-use-after-free (/usr/bin/ceph-osd+0x3c4cdbd7) in auto seastar::futurize_invoke<crimson::osd::Watch::notify_ack(unsigned long, ceph::buffer::v15_2_0::list const&)::{lambda(auto:1)#1}&, seastar::shared_ptr<crimson::osd::Notify> const&>(crimson::osd::Watch::notify_ack(unsigned long, ceph::buffer::v15_2_0::list const&)::{lambda(auto:1)#1}&, seastar::shared_ptr<crimson::osd::Notify> const&)
```

Fixes: https://tracker.ceph.com/issues/61504

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [x] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/crimson/osd/watch.cc [hunk 1]
```diff
@@ -123,13 +123,16 @@ seastar::future<> Watch::notify_ack(
...
-  ).then([this] {
-    in_progress_notifies.clear();
-    return seastar::now();
+  return seastar::do_with(shared_from_this(),
+    [this, reply_bl] (auto& this_shared) {
+    return seastar::do_for_each(in_progress_notifies,
+      [&this_shared, reply_bl] (auto notify) {
+        return notify->complete_watcher(this_shared, reply_bl);
+      }
+    ).then([this] {
```
Comment (rzarzynski) [inline src/crimson/osd/watch.cc:132]: Do we need `do_with()` and the extra level of indentation imposed by it? I'm thinking about:

```suggestion
    [this_shared=shared_from_this(), this] {
```

Comment (athanatos) [reply src/crimson/osd/watch.cc:132]: @rzarzynski's suggestion should work, I think.

File: src/crimson/osd/watch.cc [hunk 2]
```diff
@@ -300,19 +303,21 @@ void Notify::do_notify_timeout()
...
-  // to avoid use-after-free we bump up the ref counter with `guard_ptr`.
-  [[maybe_unused]] auto guard_ptr = shared_from_this();
-  for (auto& watcher : watchers) {
-    logger().debug("canceling watcher cookie={} gid={} use_count={}",
-      watcher->get_cookie(),
-      watcher->get_watcher_gid(),
-      watcher->use_count());
-    watcher->cancel_notify(ninfo.notify_id);
-  }
-  std::ignore = send_completion(std::move(watchers));
```
Comment (athanatos) [inline src/crimson/osd/watch.cc:314]: I think we can just do:

```
  std::ignore = send_completion(std::move(watchers))).then([guard_ptr] {});
```

**Review by athanatos — CHANGES_REQUESTED** [review]

File: src/crimson/osd/watch.cc (hunk 2 above)
Comment (cyx1231st) [reply src/crimson/osd/watch.cc:314]: I'm not sure whether there is any assumption about the send message ordering in this case, but it worths noting that cross-core calls cannot guarantee FIFO in the target core. With the current implementation, the `future` returned by `Connection::send()` should be chained correctly before the next message.

Also see https://github.com/ceph/ceph/pull/51147#discussion_r1206155226

**Comment by Matan-B** [comment]
From the logs, it looks like the issue is with a `crimson::osd::Notify` being deleted, while the changes here may resolve the issue, I prefer having precise logs before the next appearance. (#51852) 
```
freed by thread T1 (reactor-1) here:
    #0 0x7fd72b5af36f in operator delete(void*, unsigned long) (/lib64/libasan.so.6+0xb736f)
    #1 0x5569837a910a in std::_Rb_tree<seastar::shared_ptr<crimson::osd::Notify>, seastar::shared_ptr<crimson::osd::Notify>, std::_Identity<seastar::shared_ptr<crimson::osd::Notify> >, std::less<void>, std::allocator<seastar::shared_ptr<crimson::osd::Notify> > >::_M_erase(std::_Rb_tree_node<seastar::shared_ptr<crimson::osd::Notify> >*) (/usr/bin/ceph-osd+0x3c4eb10a)

previously allocated by thread T1 (reactor-1) here:
    #0 0x7fd72b5ae307 in operator new(unsigned long) (/lib64/libasan.so.6+0xb6307)
    #1 0x5569837b7cae in std::pair<std::_Rb_tree_iterator<seastar::shared_ptr<crimson::osd::Notify> >, bool> std::_Rb_tree<seastar::shared_ptr<crimson::osd::Notify>, seastar::shared_ptr<crimson::osd::Notify>, std::_Identity<seastar::shared_ptr<crimson::osd::Notify> >, std::less<void>, std::allocator<seastar::shared_ptr<crimson::osd::Notify> > >::_M_emplace_unique<seastar::shared_ptr<crimson::osd::Notify> >(seastar::shared_ptr<crimson::osd::Notify>&&) (/usr/bin/ceph-osd+0x3c4f9cae)
```

**Comment by Matan-B** [comment]
Closing in favor of: #51936

**Pull request closed without merging**